File: workers/research/src/black_book_research/adapters/run_health.py
"""Adapter run health evaluation mirroring domain run-health (BB-037)."""

from __future__ import annotations

from .types import EvaluateRunHealthResult, RunHealthIssue


def _is_count_within_tolerance(
    expected: int, actual: int, tolerance_fraction: float
) -> bool:
    if expected == 0:
        return actual == 0
    delta = abs(actual - expected) / expected
    return delta <= tolerance_fraction


def evaluate_run_health(
    *,
    expected_count: int,
    actual_count: int,
    expected_schema_version: str,
    observed_schema_version: str,
    count_tolerance_fraction: float = 0.15,
    null_field_rate: float | None = None,
    max_null_field_rate: float | None = None,
    missing_required_fields: tuple[str, ...] = (),
) -> EvaluateRunHealthResult:
    issues: list[RunHealthIssue] = []
    details: list[str] = []

    if not _is_count_within_tolerance(
        expected_count, actual_count, count_tolerance_fraction
    ):
        issues.append("record_count_drift")
        details.append(
            "Record count drift: "
            f"expected {expected_count}, got {actual_count} "
            f"(tolerance {count_tolerance_fraction})"
        )

    if observed_schema_version != expected_schema_version:
        issues.append("schema_version_drift")
        details.append(
            "Schema version drift: "
            f"expected {expected_schema_version}, got {observed_schema_version}"
        )

    if (
        null_field_rate is not None
        and max_null_field_rate is not None
        and null_field_rate > max_null_field_rate
    ):
        issues.append("null_field_spike")
        details.append(
            f"Null field rate {null_field_rate} exceeds max {max_null_field_rate}"
        )

    if missing_required_fields:
        issues.append("missing_required_field")
        details.append(f"Missing required fields: {', '.join(missing_required_fields)}")

    outcome = "quarantined" if issues else "success"
    return EvaluateRunHealthResult(
        outcome=outcome, issues=tuple(issues), details=tuple(details)
    )


def should_quarantine_run(result: EvaluateRunHealthResult) -> bool:
    return result.outcome == "quarantined"


def should_dead_letter_run(consecutive_quarantines: int, threshold: int = 3) -> bool:
    return consecutive_quarantines >= threshold

File: packages/domain/src/adapters/run-health.ts
/**
 * Adapter run health evaluation: record-count and schema drift → quarantine.
 */
import type { AdapterRunOutcome } from './types.js';

export const RUN_HEALTH_ISSUES = [
  'record_count_drift',
  'schema_version_drift',
  'null_field_spike',
  'missing_required_field',
] as const;

export type RunHealthIssue = (typeof RUN_HEALTH_ISSUES)[number];

export type EvaluateRunHealthInput = {
  readonly expectedCount: number;
  readonly actualCount: number;
  readonly countToleranceFraction?: number;
  readonly expectedSchemaVersion: string;
  readonly observedSchemaVersion: string;
  /** Observed null rate for a tracked field (0–1). */
  readonly nullFieldRate?: number;
  readonly maxNullFieldRate?: number;
  readonly missingRequiredFields?: readonly string[];
};

export type EvaluateRunHealthResult = {
  readonly outcome: Extract<AdapterRunOutcome, 'success' | 'quarantined'>;
  readonly issues: readonly RunHealthIssue[];
  readonly details: readonly string[];
};

function isCountWithinTolerance(
  expected: number,
  actual: number,
  toleranceFraction: number,
): boolean {
  if (expected === 0) {
    return actual === 0;
  }
  const delta = Math.abs(actual - expected) / expected;
  return delta <= toleranceFraction;
}

export function evaluateRunHealth(input: EvaluateRunHealthInput): EvaluateRunHealthResult {
  const issues: RunHealthIssue[] = [];
  const details: string[] = [];
  const tolerance = input.countToleranceFraction ?? 0.15;

  if (!isCountWithinTolerance(input.expectedCount, input.actualCount, tolerance)) {
    issues.push('record_count_drift');
    details.push(
      `Record count drift: expected ${input.expectedCount}, got ${input.actualCount} (tolerance ${tolerance})`,
    );
  }

  if (input.observedSchemaVersion !== input.expectedSchemaVersion) {
    issues.push('schema_version_drift');
    details.push(
      `Schema version drift: expected ${input.expectedSchemaVersion}, got ${input.observedSchemaVersion}`,
    );
  }

  if (
    input.nullFieldRate !== undefined &&
    input.maxNullFieldRate !== undefined &&
    input.nullFieldRate > input.maxNullFieldRate
  ) {
    issues.push('null_field_spike');
    details.push(
      `Null field rate ${input.nullFieldRate} exceeds max ${input.maxNullFieldRate}`,
    );
  }

  if (input.missingRequiredFields?.length) {
    issues.push('missing_required_field');
    details.push(`Missing required fields: ${input.missingRequiredFields.join(', ')}`);
  }

  return {
    outcome: issues.length ? 'quarantined' : 'success',
    issues,
    details,
  };
}

export function shouldQuarantineRun(result: EvaluateRunHealthResult): boolean {
  return result.outcome === 'quarantined';
}

export function shouldDeadLetterRun(
  consecutiveQuarantines: number,
  threshold = 3,
): boolean {
  return consecutiveQuarantines >= threshold;
}
